### PR TITLE
Add webconfig to roleblock

### PR DESCRIPTION
**Pull request by pwalczysko**
@sbesson this PR moves the cofig block directly under the web role, as discussed. Copied the block from from the ``learning.yml`` playbook, deleted everything which I did not recognize as useful or know, and added all the config as per the ``omero-web-outreach-webapps.omero`` file listings.

Nevertheless, the playbook with these changes still fails with

```
fatal: [ome-training-3.openmicroscopy.org]: FAILED! => {"changed": false, "msg": "Unable to start service omero-web: Job for omero-web.service failed because the control process exited with error code. See \"systemctl status omero-web.service\" and \"journalctl -xe\" for details.\n"}
```

### CODE REVIEW

File: omero/training-server/playbook.yml [hunk 1]
```diff
@@ -76,6 +76,43 @@
...
+          - ["Figure", "figure_index", {"title": "Open Figure in new tab", "target": "_blank"}]
+          - ["Tag Search", "tagsearch"]
+          - ["Genes", {"query_string": {"experimenter": -1}, "viewname": "maprindex_gene"}, {"title": "Find Gene annotations"}]
+          - ["Key-Value", {"viewname": "maprindex_keyvalue"}, {"title": "Search for manually-added Key-Value pairs"}]
+        omero.web.open_with:
+          - ["omero_fpbioimage", "fpbioimage_index", {"supported_objects":["image"], "script_url": "fpbioimage/openwith.js", "label": "FPBioimage"}]
+          - ["omero_iviewer", "omero_iviewer_index", {"supported_objects":["images", "dataset", "well"], "script_url": "omero_iviewer/openwith.js", "label": "OMERO.iviewer"}]
+          - ["omero_figure", "new_figure", {"supported_objects":["images"], "target": "_blank", "label": "OMERO.figure"}]
+        omero.web.apps:
+          - "omero_gallery"
```
Comment (sbesson) [inline omero/training-server/playbook.yml:94]:
```
[sbesson@ome-training-3 ~]$ sudo service omero-web status
...
Mar 18 12:29:17 ome-training-3 omero[14819]: ModuleNotFoundError: No module named 'omero_gallery'
```

But `omero-gallery` is not installed?

**Comment by pwalczysko** [comment]
```
...
TASK [ome.omero_web : omero web | configuration 00-omero-web.omero] ************
--- before: /opt/omero/web/config/00-omero-web.omero
+++ after: /Users/pwalczysko/.ansible/tmp/ansible-local-90782m0302681/tmpr53bxzae/00-omero-web-omero.j2
@@ -3,7 +3,6 @@
 config drop default
 
 # Additional custom options
-config set -- omero.web.apps '["omero_iviewer", "omero_figure", "omero_fpbioimage", "omero_webtagging_autotag", "omero_webtagging_tagsearch", "omero_parade", "omero_mapr"]'
 config set -- omero.web.mapr.config '[[{"menu": "gene", "config": {"default": ["Gene Symbol"], "all": ["Gene Symbol", "Gene Identifier"], "ns": ["openmicroscopy.org/mapr/gene"], "label": "Gene"}}, {"menu": "keyvalue", "config": {"default": ["Any Value"], "all": [], "ns": ["openmicroscopy.org/omero/client/mapAnnotation"], "label": "KeyValue"}}]]'
 config set -- omero.web.open_with '[["omero_fpbioimage", "fpbioimage_index", {"supported_objects": ["image"], "script_url": "fpbioimage/openwith.js", "label": "FPBioimage"}], ["omero_iviewer", "omero_iviewer_index", {"supported_objects": ["images", "dataset", "well"], "script_url": "omero_iviewer/openwith.js", "label": "OMERO.iviewer"}], ["omero_figure", "new_figure", {"supported_objects": ["images"], "target": "_blank", "label": "OMERO.figure"}]]'
 config set -- omero.web.ui.center_plugins '[["Parade", "omero_parade/init.js.html", "omero_parade"], ["Auto Tag", "omero_webtagging_autotag/auto_tag_init.js.html", "auto_tag_panel"]]'

changed: [ome-training-3.openmicroscopy.org]
...
```

**Comment by pwalczysko** [comment]
with https://github.com/ome/prod-playbooks/pull/334/commits/276ed12a4f40215416e2b6fe3974e02379e3a6eb I have a successfull wet run of the playbook on ome-training-3. Web restarts and there are no errors. But the output bit of https://github.com/ome/prod-playbooks/pull/334#issuecomment-1072502699 is suggesting that the ``....omero`` file is replaced by a ``...j2`` file as a source of config ? This might explain the duplication of the top links, the central pane links and of the open with links observed on ome-training-3. Also, the "Data" top link is gone, see screenshot.


ome-training-3
Screenshot 2022-03-18 at 15 22 51
Screenshot 2022-03-18 at 15 19 12

outreach (without this PR)

Screenshot 2022-03-18 at 15 19 42


cc @will-moore

**Comment by pwalczysko** [comment]
not sure what is wrong. The top links are as expected. But they do not work, maybe I am not installing the apps ?

**Comment by pwalczysko** [comment]
I think https://github.com/ome/prod-playbooks/pull/334/commits/024f1c243d2bdfa734ff98e16708248d8359ede7 should fix the apps installation - did not have them in the ``config set`` block, so it could not have worked.
Will be able to verify once my ssh access is restored, see PRs on mngmt tools repo.

**Comment by pwalczysko** [comment]
@sbesson with the latest commit fixing the formatting of mapr and extra config, deployed on ``ome-training-3``. Checked there and all works. The difference between the existing and the new config seems to me to be consisting only in the ordering of the apps, which should not bear any importance ?

```
diff -ur new.config outreach.config 
--- new.config	2022-03-21 13:12:02.837419621 +0000
+++ outreach.config	2022-03-21 13:12:12.488465336 +0000
@@ -1,4 +1,4 @@
-omero.web.apps=["omero_iviewer", "omero_figure", "omero_fpbioimage", "omero_webtagging_autotag", "omero_webtagging_tagsearch", "omero_parade", "omero_mapr", "django_prometheus"]
+omero.web.apps=["django_prometheus", "omero_figure", "omero_fpbioimage", "omero_iviewer", "omero_webtagging_autotag", "omero_webtagging_tagsearch", "omero_parade", "omero_mapr"]
 omero.web.mapr.config=[{"menu": "gene", "config": {"default": ["Gene Symbol"], "all": ["Gene Symbol", "Gene Identifier"], "ns": ["openmicroscopy.org/mapr/gene"], "label": "Gene"}}, {"menu": "keyvalue", "config": {"default": ["Any Value"], "all": [], "ns": ["openmicroscopy.org/omero/client/mapAnnotation"], "label": "KeyValue"}}]
 omero.web.middleware=[{"index": 1, "class": "django.middleware.common.BrokenLinkEmailsMiddleware"}, {"index": 2, "class": "django.middleware.common.CommonMiddleware"}, {"index": 3, "class": "django.contrib.sessions.middleware.SessionMiddleware"}, {"index": 4, "class": "django.middleware.csrf.CsrfViewMiddleware"}, {"index": 5, "class": "django.contrib.messages.middleware.MessageMiddleware"}, {"index": 6, "class": "django.middleware.clickjacking.XFrameOptionsMiddleware"}, {"index": 0, "class": "django_prometheus.middleware.PrometheusBeforeMiddleware"}, {"index": 1000, "class": "django_prometheus.middleware.PrometheusAfterMiddleware"}]
 omero.web.nginx_server_extra_config=["include /etc/nginx/conf.d-nested-includes/*.conf;", "location = /omero-ws {\n  proxy_pass https://127.0.0.1:4066;\n  proxy_http_version 1.1;\n  proxy_set_header Upgrade $http_upgrade;\n  proxy_set_header Connection $connection_upgrade;\n  proxy_read_timeout 86400;\n}\n"]
```

**Comment by pwalczysko** [comment]
puzzled by the molecule failures here - did not touch the notebooks pertaining to ome-dundeeomero and simlar... ?

**Comment by sbesson** [comment]
There are two types of failures here:


- the `ome-pg-prod` playbook fails with

```
    fatal: [ome-pg-prod1.openmicroscopy.org]: FAILED! => {"changed": false, "msg": "No package matching 'postgresql96' found available, installed or updated", "rc": 126, "results": ["No package matching 'postgresql96' found available, installed or updated"]}
```

which is related to https://github.com/ome/ansible-role-postgresql/pull/26 i.e. PSQL 9.6 is EOL and has been removed from the CentOS packages. 

-  the `ome-dundeeomero`, `omero-training` and `nightshade-webclients` tests all fail while linting the training server playbook with

```
 WARNING  Listing 1 violation(s) that are fatal
    [204] Lines should be no longer than 160 chars
    omero/training-server/playbook.yml:104
              - ["omero_iviewer", "omero_iviewer_index", {"supported_objects":["images", "dataset", "well"], "script_url": "omero_iviewer/openwith.js", "label": "OMERO.iviewer"}]
    
Warning: [E204] Lines should be no longer than 160 chars
```


I can handle the former. Do you want to look into the latter?  I assume splitting the one-line string into a YML representation similar to 0fbf81c3bffa33288c19213a01e486f0258f73e7 would be an option

**Comment by pwalczysko** [comment]
> I can handle the former. Do you want to look into the latter? I assume splitting the one-line string into a YML representation similar to https://github.com/ome/prod-playbooks/commit/0fbf81c3bffa33288c19213a01e486f0258f73e7 would be an option

unlike in the mapr config case, the iviewer item in the Open with list does not have nice headers everywhere. The list just starts with something which looks like a value of some (unnamed?) parameter

```
- ["omero_iviewer", "omero_iviewer_index", {"supported_objects":["images", "dataset", "well"], "script_url": "omero_iviewer/openwith.js", "label": "OMERO.iviewer"}]
```

I tried to approximate with

```
- omero_iviewer
          - omero_iviewer_index
          - supported_objects:
              - "images"
              - "dataset"
              - "well"
          - script_url: 
              - "omero_iviewer/openwith.js"
          - label: "OMERO.iviewer"
 ```

which does not fail in ansible, but is not delivering the final result, because, obviously, the overarching syntax which would suggest "this is iviewer open with item" is not there.
How to solve and create a nice "pen" for iviewer ?

**Comment by sbesson** [comment]
I think you're almost there. Using https://onlineyamltools.com/convert-json-to-yaml, I got

```yaml
- omero_iviewer
- omero_iviewer_index
- supported_objects:
    - images
    - dataset
    - well
  script_url: omero_iviewer/openwith.js
  label: OMERO.iviewer
```

**Comment by pwalczysko** [comment]
> I think you're almost there. Using https://onlineyamltools.com/convert-json-to-yaml, I got
> 
> ```yaml
> - omero_iviewer
> - omero_iviewer_index
> - supported_objects:
>     - images
>     - dataset
>     - well
>   script_url: omero_iviewer/openwith.js
>   label: OMERO.iviewer
> ```

Thank you @sbesson - after some trials and errors, I realized there must be, surprisingly, a double dash as per https://github.com/ome/prod-playbooks/pull/334/commits/31b3da186153c42f902697e047b33a2c1588d157 , otherwise the web fails with an error complaining about open with and iviewer. The comparison of the before and after showed that one encompassing square bracket is missing around the whole "open with iviewer" bit. Hence the additional starting dash. Tested on ome-training-3, seems to work fine.

**Comment by pwalczysko** [comment]
@sbesson this https://github.com/ome/prod-playbooks/pull/334/commits/31b3da186153c42f902697e047b33a2c1588d157 seems to have solved the milecule failures as well, except the ``ome-pg-prod``.

**Comment by pwalczysko** [comment]
Edit: deployed successfully also on ``ome-training-4`` As with the ``ome-training-3``, it was necessary and sufficient to 

delete the file
``/opt/omero/web/config/omero-web-outreach-webapps.omero ``

and run the playbook here.

**Review by sbesson — COMMENTED** [review]
The issue with the training server is

```
    fatal: [ome-outreach]: FAILED! => {"changed": false, "msg": "Could not find the requested service nginx: host"}
```

which I think I traced back to a config change in this PR.

For `ome-pg-prod`, I think we'll need to exclude the Molecule test for now, I am looking for the less invasive way to do this and capture as an issue

**Comment by pwalczysko** [comment]
@sbesson your commit is added. The Molecule does not seem to start ?

**Comment by sbesson** [comment]
> @sbesson your commit is added. The Molecule does not seem to start ?

I think that's related to https://www.githubstatus.com/incidents/83lq7ftk19r5 rather than my commit ;)

**Comment by pwalczysko** [comment]
``docker-prod`` is failing now

**Review by sbesson — APPROVED** [review]
Very nice. Happy to get that in assuming you are happy about the behavior. We can roll the change out to all training servers and then focus on #335.

Another cleanup that can be carried out here or in #335 is to remove the outdated `omero-web-outreach-webapps.omero`

**Comment by pwalczysko** [comment]
Thank you @sbesson , merging.

**Pull request merged**